### DISCUSSION TITLE
REST interface setup questions

### DISCUSSION THREAD

**Discussion by guykatz** [microstream-one/microstream#549] [Q&A] [answered]
Hi all;
trying to integrade REST for microstream into wildfly...
using the StorageRestService implementation of spark-java brings with it a whole lot of dependencies. some light such as GSON, some on the heavier side such as jetty and websocket jars.
1. per the documentation here: https://docs.microstream.one/manual/storage/rest-interface/setup.html it states there spark-java is one of the implementations. are there other implementations? i do not see another microstream-storage-restservice-XXX.jar in the REST docuemntation, other than the spark-java, that suggests another existing implementation...
2. we are using a jakarta EE containers that uses a fork of tomcat as its webserver, its seems wierd that I should bring jetty to accomodate the rest endpoints of microstream... is there an alternative?
thanks!

**Comment by hg-ms** [comment]
Hello,
The spark-java REST service was intended to provide the Api to standalone java apps too, that’s why it has jetty and other stuff included. There is currently no other implementation.
If you like to get rid of all those dependencies, maybe you can implement the four required routes directly within your app using your preferred frameworks.

**Comment by guykatz** [comment]
10x @hg-ms 
I would love to implement it. Exposing rest is easy for me, I am just not sure how to hook it to the adapter or other needed REST related layer microstream already provided.
Can you give me some pointers?

**Comment by hg-ms** [comment, marked as answer]
I’ll gladly give you some hints.

1.	The `one.microstream.storage.restadapter.types.StorageRestAdapter` provides the Java API to access the internal Microstream data from the rest service.
It requires an already running `EmbeddedStorageManager`, to create it use the New method:
` one.microstream.storage.restadapter.types.StorageRestAdapter.New(StorageManager)`
It provides the four method needed to implement the four REST calls to satisfy the viewer app:
- getObject
- getUserRoot
- getTypeDictionary (optional, not used by viewer app)
- getStorageFilesStatistics
2.	As long as you don’t want to implement an own viewer you need to name the Routes accordingly, the viewer expects:
"dictionary",
"root",
"object",
"maintenance/filesStatistics"
See `StorageRestClientJersey` as reference.

3.	For the url query parameters, have a look at `one.microstream.storage.restservice.sparkjava.types.RouteGetObject` for an example.

4.	The Viewer expects to get JSON data, that’s why we used GSON. There is one special handling for ` java.util.Date`, please see ` one.microstream.storage.restservice.sparkjava.types.StorageViewDataConverterJson`.

5.	We use Java SPI in our implementation but that is not necessary.

I’m sure I missed some things, so don’t hesitate if you have any questions.

Best regards

**Reply by guykatz** [reply]
thanks @hg-ms 
thats a good amount of info for me to start...
I will take my time and post further questions if needed.
thanks again.

**Reply by guykatz** [reply]
@hg-ms does the current gui rest client support any kind of authentication? my REST interfaces are all protected with HTTP basic auth as a minimum...
the UI just shows a URL to connect to without user pass etc...
thanks

**Reply by hg-ms** [reply]
There is no authentication support in the viewer.

**Reply by guykatz** [reply]
basic auth for REST App: https://github.com/microstream-one/microstream/compare/master...guykatz:microstream:master
this is an example but working one.
can be used by anyone who needs http basic auth. user and password are sent as system properties and not via UI because I dont know vaadin
only a small change in MS code and a new small class for basic auth utilities as a REST client filter needed...

**Comment by guykatz** [comment]
@hg-ms 
StirageRestAdapter getObject uses long values as input parameters.
however, the documentation here: https://docs.microstream.one/manual/storage/rest-interface/rest-api.html#_get_object shows we need to use Int types in the query parameters. 
my microstream first id is: 1000000000000000028 which I think does not fit a Java Interger.
is this a documentation mistake? 
thanks 

UPDATE: after chaging the jaxrs input param from integers to longs all started to work fine.
thanks

**Reply by hg-ms** [reply]
Thanks for that finding, the getObject call indeed expects long values as parameters, need to fix that in the Docu.
